core/src/main/scala/org/apache/spark/util/collection/OpenHashSet.scala

```diff
@@ -110,6 +110,18 @@ class OpenHashSet[@specialized(Long, Int, Double, Float) T: ClassTag](
     this
   }
 
+  /**
+   * Check if a key exists at the provided position using object equality rather than
+   * cooperative equality. Otherwise, hash sets will mishandle values for which `==`
+   * and `equals` return different results, like 0.0/-0.0 and NaN/NaN.
```

Comment:
I was told that in scala `==` is the same as `equals`, but `eq` is a different operator. I need to refresh my knowledge now :)

Reply:
Yes, the differences are subtle:

```scala
scala> 0.0 == -0.0
val res0: Boolean = true

scala> 0.0 equals -0.0
val res1: Boolean = false

scala> Double.NaN == Double.NaN
val res2: Boolean = false

scala> Double.NaN equals Double.NaN
val res3: Boolean = true
```

There is a long discussion on the Scala forums from 2017 about this difference and some of the problems it causes:

[Can we get rid of cooperative equality?](https://contributors.scala-lang.org/t/can-we-get-rid-of-cooperative-equality/1131)